designs/pluggable-message-routing/README.md

```diff
@@ -0,0 +1,263 @@
+
+**At this stage, the life-cycle hooks approach is the preferred model.**
+
+#### Stack of layers
+
+The working assumption for this feature has always been to take an approach similar
+to that of Express middleware:
+
+Each message is passed to the top layer of the stack. That layer can do whatever
+it wants with the message before either disposing of it or passing it on to the
+next layer in the stack.
+
+Node-RED would provide a core `LocalRouter` layer that knows how to route messages
+to nodes in the local runtime.
+
+The stack layers could be customised via either the settings file or via runtime API.
+
+The problem with this model comes when you have unrelated things wanting to customise
+the stack.
+
+For example, if a user provided a custom layer to add additional logging, they may
+specify a stack that looks like:
+```
+1. MyCustomLogger
+2. RED.router.LocalRouter
+```
+
+They then enable the Flow Debugger that has to insert itself into this stack. That would
+mean the Flow Debugger would have to keep track of the old stack configuration and be
+able to restore it when the Debugger is disabled.
+
+This could get very complicated to manage.
+
+
+#### Life-Cycle Hooks
+
+This is similar to the Fastify model of registering hooks at certain key points
+in the life-cycle of a message.
+
+Rather than leave the stack as being completely customisable, we recognise there
+is a core set of steps that every message has to go through. Within those steps
+are points where some custom code *might* want to run.
+
+The following diagram shows a possible set of hook points. This list may well get
+refined as the design progresses. The orange lines show the span of synchronous calls.
+
+![](message-router-events.png)
+
+1. `preSend` - passed an array of `SendEvent` objects. The messages inside these objects
+   are exactly what the node has passed to `node.send` - meaning there could be duplicate
+   references to the same message object.
+
+2. `preRoute` - called once for each SendEvent object in turn
+
+3. `onSend` - the local router has identified the node it is going to send to. At
+   this point, the message has been cloned if needed.
+
+4. `postSend` - the message has been dispatched to be delivered asynchronously
+(unless the sync delivery flag is set, in which case it would be continue as synchronous delivery)
+
+5. `onReceive` - a node is about to receive a message
+
+6. `postReceive` - the message has been passed to the node's `input` handler
+
+7. `onDone`, `onError` - the node has completed with a message or logged an error
+
+
+*The exact function signature for each of these hooks needs to be designed.*
+
+They operate on `SendEvent` objects. Each objects contains the message and metadata
+needed through the process.
+
+They start out looking like:
+
+```json
+{
+    "msg": "<message object>",
+    "source": {
+        "id": "<node-id>",
+        "node": "<node-object>",
+        "port": "<index of port being sent on>",
```

Comment:
what does port mean in this instance ?

Reply:
It identifies which output port (`0`-indexed) the message was sent.

Reply:
So if on port 1 or above it is already cloned ?

Reply:
Any built-in cloning that happens will occur between the `preRoute` and `onSend` stages - and will only occur if the `cloneMessage` property is `true`. The port number has no direct bearing on the logic here.

For example, if a node sends a single message out of port 2 and there's only one wire connected to that port, then the node will set `cloneMessage` to `false` because it needs no further cloning.